File: Coding Interview Patterns (Alex Xu)/Hash Maps and Sets/Geometric Sequence Triplets/GeometricSequenceTriplets.java
import java.util.HashMap;

public class GeometricSequenceTriplets {
    public static void main(String[] args) {
        // Original sample test case:
        System.out.println(solution(new int[] { 2, 1, 2, 4, 8, 8 }, 2)); // Expected output: 5

        // Edge Case 1: Empty Array
        System.out.println(solution(new int[] {}, 2)); // Expected output: 0

        // Edge Case 2: Single Element Array
        System.out.println(solution(new int[] { 1 }, 2)); // Expected output: 0

        // Edge Case 3: Two Elements Only
        System.out.println(solution(new int[] { 1, 2 }, 2)); // Expected output: 0

        // Edge Case 4: All Identical Elements with r = 1
        System.out.println(solution(new int[] { 1, 1, 1, 1 }, 1)); // Expected output: 4

        // Edge Case 5: Larger Set of Identical Elements with r = 1
        System.out.println(solution(new int[] { 1, 1, 1, 1, 1 }, 1)); // Expected output: 10

        // Edge Case 6: No Valid Triplets (Different Ratio)
        System.out.println(solution(new int[] { 1, 2, 3, 4, 5 }, 3)); // Expected output: 0

        // Edge Case 7: Valid Triplets with Mixed Order and Duplicates
        System.out.println(solution(new int[] { 1, 3, 9, 9, 27, 81 }, 3)); // Expected output: 6

        // Edge Case 8: Case with r = 0
        // Here, we catch and print the exception.
        try {
            System.out.println(solution(new int[] { 0, 0, 0 }, 0));
        } catch (Exception e) {
            System.out.println("Error for r=0: " + e);
        }

        // Edge Case 9: Negative Numbers with a Negative Ratio
        System.out.println(solution(new int[] { 1, -1, 1, -1, 1 }, -1)); // Expected output: Depends on array order

        // Edge Case 10: Duplicates in a Non-r = 1 Geometric Progression
        System.out.println(solution(new int[] { 2, 2, 2, 4, 4, 8 }, 2)); // Expected output: 6

    }

    public static int solution(int[] inputArr, int commonRatio) {
        HashMap<Integer, Integer> leftMap = new HashMap<>();
        HashMap<Integer, Integer> rightMap = new HashMap<>();

        // can use 'getOrDefault' built-in
        // rightMap.put(x, rightMap.getOrDefault(x, 0) + 1);

        for (int currentNum : inputArr) {
            if (!rightMap.containsKey(currentNum)) {
                rightMap.put(currentNum, 1);
                leftMap.put(currentNum, 0);
            } else {
                rightMap.put(currentNum, rightMap.get(currentNum) + 1);
            }
        }

        int count = 0;

        // I think int x is automatically converted to Integer while doing
        // HashMap.get(x) or HashMap.put(x, int)
        // because of keys of HashMaps are Integers
        for (int x : inputArr) {

            rightMap.put(x, rightMap.get(x) - 1);

            if (x % commonRatio == 0) {
                if (leftMap.containsKey(x / commonRatio) && rightMap.containsKey(x * commonRatio)) {
                    count += leftMap.get(x / commonRatio) * rightMap.get(x * commonRatio);
                }
            }

            leftMap.put(x, leftMap.get(x) + 1);

        }

        return count;
    }
}

File: Coding Interview Patterns (Alex Xu)/Hash Maps and Sets/Geometric Sequence Triplets/geometric_sequence_triplets.py
def solution(inputArr, r):
    leftMap = {}
    rightMap = {}
    for i in inputArr:
        # you can use 'defaultdict' to ensure the default value of 0 is returned when accessing a key that doesn’t exist in the hash map.
        # I prefer doing manually
        if i not in rightMap.keys():
            rightMap[i] = 1
            # not to get index when leftMap[x//r] for new ints
            leftMap[i] = 0
        else:
            rightMap[i] += 1
    count = 0
    for x in inputArr:
        # remove current int from right
        # technically, removing x before count doesn't matter because finding x*r in rightMap will never count x (unless r is 1)
        # but conceptually, current number should not be included inside the array right side of the number
        rightMap[x] -= 1

        if x % r == 0:
            if x//r in leftMap.keys() and x*r in rightMap.keys():
                count += leftMap[x//r] * rightMap[x*r]

        # add current int to the left
        leftMap[x] += 1
            
    return count

def main():
    # Original sample test case:
    print(solution([2, 1, 2, 4, 8, 8], 2))  # Expected output: 5

    # Edge Case 1: Empty Array
    print(solution([], 2))  # Expected output: 0

    # Edge Case 2: Single Element Array
    print(solution([1], 2))  # Expected output: 0

    # Edge Case 3: Two Elements Only
    print(solution([1, 2], 2))  # Expected output: 0

    # Edge Case 4: All Identical Elements with r = 1
    print(solution([1, 1, 1, 1], 1))  # Expected output: 4

    # Edge Case 5: Larger Set of Identical Elements with r = 1
    print(solution([1, 1, 1, 1, 1], 1))  # Expected output: 10

    # Edge Case 6: No Valid Triplets (Different Ratio)
    print(solution([1, 2, 3, 4, 5], 3))  # Expected output: 0

    # Edge Case 7: Valid Triplets with Mixed Order and Duplicates
    print(solution([1, 3, 9, 9, 27, 81], 3))  # Expected output: 6

    # Edge Case 8: Case with r = 0
    # Here, we catch and print the exception.
    try:
        print(solution([0, 0, 0], 0))
    except Exception as e:
        print("Error for r=0:", e)

    # Edge Case 9: Negative Numbers with a Negative Ratio
    print(solution([1, -1, 1, -1, 1], -1))  # Expected output: Depends on array order

    # Edge Case 10: Duplicates in a Non-r = 1 Geometric Progression
    print(solution([2, 2, 2, 4, 4, 8], 2))  # Expected output: 6
    
main()